### PR TITLE
Fetch and populate source data for consumer records

### PR DESCRIPTION
**Pull request by acco**
This pull request enhances the functionality of the `Consumers` module by implementing source data fetching for consumer records. The main changes include:

1. Added `put_source_data/3` function to fetch and populate source data for consumer records.
2. Implemented `fetch_records_data/3` to retrieve data from the source table for a group of records.
3. Added support for handling both single and compound primary keys when fetching source data.
4. Introduced type casting for different column types to ensure proper data representation.
5. Updated `receive_for_consumer/2` to include source data fetching for record-type messages.
6. Added helper functions in the `Postgres` module for type conversion and parameter generation.
7. Expanded test coverage to include various scenarios for source data fetching, including edge cases and error handling.

These changes improve the consumer record processing by providing more complete and accurately typed data from the source tables.

### CODE REVIEW

**Comment by acco** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/122?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#122** <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/122?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#121** <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/121?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#120** <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/120?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#119** <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/119?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#118** <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/118?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#117** <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/117?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#116** <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/116?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#115** <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/115?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#114** <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/114?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#113** <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/113?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#112** <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/112?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#111** <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/111?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#110** <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/110?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @acco and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: lib/sequin/consumers/consumers.ex [hunk 1]
```diff
@@ -469,7 +475,121 @@ defmodule Sequin.Consumers do
...
+        # which is too challenging to pull off with Ecto fragments
+        pk_column_names = pk_columns |> Enum.map(& &1.name) |> Enum.map_join(", ", &Postgres.quote_name/1)
+
+        value_params =
+          Enum.map_join(1..record_count, ", ", fn n ->
+            Postgres.parameterized_tuple(pk_column_count, (n - 1) * pk_column_count)
+          end)
+
+        "where (#{pk_column_names}) in (#{value_params})"
+      end
```
Comment (acco) [inline lib/sequin/consumers/consumers.ex:546]: It weirds me out that we need to fork here, but apparently we can't use single-element tuples in the first clause

File: lib/sequin/consumers/consumers.ex [hunk 2]
```diff
@@ -469,7 +475,121 @@ defmodule Sequin.Consumers do
...
+        |> Enum.zip(pk_types)
+        |> Enum.map(fn {value, type} -> cast_value(value, type) end)
+      end)
+      |> List.flatten()
+
+    where_clause =
+      if pk_column_count == 1 do
+        # This one needs to not use a row tuple on the left- or right-hand sides
+        # pk_column_name = Postgres.quote_name(List.first(pk_columns).name)
+        pk_column_name = "id"
```
Comment (acco) [inline lib/sequin/consumers/consumers.ex:533]: whoops, debugging crept through

File: lib/sequin/consumers/consumers.ex [hunk 3]
```diff
@@ -469,7 +475,121 @@ defmodule Sequin.Consumers do
...
+        "where #{pk_column_name} in #{Postgres.parameterized_tuple(record_count)}"
+      else
+        # the where clause is (col1, col2) IN ((val1, val2), (val3, val4))
+        # which is too challenging to pull off with Ecto fragments
+        pk_column_names = pk_columns |> Enum.map(& &1.name) |> Enum.map_join(", ", &Postgres.quote_name/1)
```
Comment (acco) [inline lib/sequin/consumers/consumers.ex:538]: We could do it with a macro...

File: lib/sequin/consumers/consumers.ex [hunk 4]
```diff
@@ -469,7 +475,121 @@ defmodule Sequin.Consumers do
...
+      # Match the results with the original records
+      updated_records =
+        records
+        |> Enum.zip(result.rows)
+        |> Enum.map(fn {record, row} ->
+          data =
+            table.columns
+            |> Enum.zip(row)
+            |> Map.new(fn {col, val} ->
+              casted_val = if col.type == "uuid", do: UUID.binary_to_string!(val), else: val
```
Comment (acco) [inline lib/sequin/consumers/consumers.ex:561]: I'm sure there's others we'll need to cast, will test

**Comment by RTLS** [comment]
> This pull request enhances the functionality of the `Consumers` module by implementing source data fetching for consumer records. The main changes include:
> 
> 1. Added `put_source_data/3` function to fetch and populate source data for consumer records.
> 2. Implemented `fetch_records_data/3` to retrieve data from the source table for a group of records.
> 3. Added support for handling both single and compound primary keys when fetching source data.
> 4. Introduced type casting for different column types to ensure proper data representation.
> 5. Updated `receive_for_consumer/2` to include source data fetching for record-type messages.
> 6. Added helper functions in the `Postgres` module for type conversion and parameter generation.
> 7. Expanded test coverage to include various scenarios for source data fetching, including edge cases and error handling.
> 
> These changes improve the consumer record processing by providing more complete and accurately typed data from the source tables.

These are very helpful as an overview before I dive into the review!

File: lib/sequin/consumers/consumers.ex (hunk 1 above)
Comment (RTLS) [reply lib/sequin/consumers/consumers.ex:546]: a bit strange yes but fine

File: lib/sequin/consumers/consumers.ex (hunk 3 above)
Comment (RTLS) [reply lib/sequin/consumers/consumers.ex:538]: don't temp me !

File: lib/sequin/consumers/consumers.ex [hunk 5]
```diff
@@ -469,7 +475,121 @@ defmodule Sequin.Consumers do
...
+      end
+
+    query = "select * from #{Postgres.quote_name(table.schema, table.name)} #{where_clause}"
+
+    # Execute the query
+    with {:ok, result} = Postgrex.query(conn, query, casted_pks) do
+      # Match the results with the original records
+      updated_records =
+        records
+        |> Enum.zip(result.rows)
```
Comment (RTLS) [inline lib/sequin/consumers/consumers.ex:555]: I'm not convinced this zip is safe- we need to guarantee the ordering of records and result.rows is the same

```
The assumption that records and result.rows are in the same order and can be safely zipped together is not guaranteed to be safe. Here's why:

Order of query results: PostgreSQL does not guarantee a specific order of results unless an ORDER BY clause is used in the query. The current query doesn't include an ORDER BY clause.
IN clause behavior: When using an IN clause (or its row-value expression equivalent), the database is free to process these values in any order it finds most efficient.
Parallel query execution: Depending on the PostgreSQL configuration and query plan, parts of the query might be executed in parallel, which could affect the order of results.
```

**Review by RTLS — APPROVED** [review]

File: lib/sequin/consumers/consumers.ex (hunk 5 above)
Comment (acco) [reply lib/sequin/consumers/consumers.ex:555]: Good catch will fix
